src/ATen/native/xpu/sycl/LossNLL2dKernels.cpp

```diff
@@ -0,0 +1,569 @@
+#pragma clang diagnostic push
+#pragma GCC diagnostic push
+// Avoid SYCL compiler return-type error
+#pragma clang diagnostic ignored "-Wreturn-type"
+#pragma GCC diagnostic ignored "-Wreturn-type"
+#include <ATen/ATen.h>
+#include <ATen/core/TensorAccessor.h>
+#include <ATen/native/CanUse32BitIndexMath.h>
+#include <ATen/native/Resize.h>
+#include <ATen/native/xpu/sycl/Atomics.h>
+#include <ATen/native/xpu/sycl/GroupReduceUtils.h>
+#include <comm/SYCLContext.h>
+
+namespace at::native::xpu {
+inline Tensor optional_contiguous(const Tensor& source) {
+  return source.defined() ? source.contiguous() : source;
+}
+
+template <typename scalar_t>
+inline const scalar_t* optional_data(const Tensor& source) {
+  return source.defined() ? source.const_data_ptr<scalar_t>() : nullptr;
+}
+
+void check_inputs_nll_loss2d(
+    const Tensor& input,
+    const Tensor& target,
+    const Tensor& weight) {
+  TORCH_CHECK(
+      target.dim() == 3,
+      "only batches of spatial targets supported (3D tensors)"
+      " but got targets of size: : ",
+      target.sizes());
+  TORCH_CHECK(
+      input.dim() == 4,
+      "only batches of spatial inputs supported (4D tensors), "
+      "but got input of size: ",
+      input.sizes());
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == input.size(1),
+      "weight tensor should be defined either for all or no classes");
+
+  TORCH_CHECK(
+      input.size(0) == target.size(0) && input.size(2) == target.size(1) &&
+          input.size(3) == target.size(2),
+      "input and target batch or spatial sizes don't match: target ",
+      target.sizes(),
+      ", input ",
+      input.sizes());
+}
+
```

Comment:
It should be operator level logic. And CPU does it in operator level. Even we are following CUDA implementation, we still need evaluate whether CUDA impl is reasonable. Let's do right thing to move them to operator level.